File: scripts/validate-family.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { Family } from './family-model.mjs';
import { clean, exists, gitText, readToml } from './family-lib.mjs';

function validateDependencies(file, root) {
  const queue = [readToml(file)];
  while (queue.length) {
    for (const value of Object.values(queue.pop())) {
      if (!value || typeof value !== 'object' || Array.isArray(value)) continue;
      queue.push(value);
      if (value.git && (value.branch || !(value.tag || value.rev))) throw new Error(`${file}: Git dependency needs an immutable pin`);
      if (value.path) {
        const relative = path.relative(root, path.resolve(path.dirname(file), value.path));
        if (relative.startsWith('../') || path.isAbsolute(relative)) throw new Error(`${file}: committed cross-repo path dependency`);
      }
    }
  }
}
function validateTree(root) {
  for (const relative of gitText(root, 'ls-files', '-z').split('\0')) {
    const file = path.join(root, relative);
    if (!exists(file) || !fs.statSync(file).isFile() || relative.startsWith('conformance/fixtures/')) continue;
    if (path.basename(file) === 'Cargo.toml') validateDependencies(file, root);
    if (relative.startsWith('.github/workflows/') || relative === 'action.yml') {
      for (const [, use] of fs.readFileSync(file, 'utf8').matchAll(/\buses:\s*["']?([^\s"'#]+)/g)) {
        if (!use.startsWith('./') && !/^[^@]+@[a-f0-9]{40}$/.test(use)) throw new Error(`${file}: unpinned action ${use}`);
      }
    }
  }
}
try {
  const family = new Family(path.dirname(path.dirname(fileURLToPath(import.meta.url))));
  const checkouts = process.argv.includes('--checkouts');
  readToml(path.join(family.hub, 'Cargo.lock'));
  for (const repo of family.repos) {
    if (!family.existing(repo)) {
      if (checkouts) throw new Error(`missing component: ${repo.name}`);
      continue;
    }
    const directory = family.path(repo);
    validateTree(directory);
    for (const required of ['AGENTS.md', 'SPLIT.md', 'agent/owner-map.json', 'agent/test-map.json', 'agent/generated-zones.toml', 'scripts/ci-local.sh', 'ops/ci/required.sh']) {
      if (!exists(path.join(directory, required))) throw new Error(`${repo.name}: missing ${required}`);
    }
    const pin = family.pins.get(repo.name);
    if (checkouts && pin) {
      clean(directory);
      if (gitText(directory, 'rev-parse', 'HEAD') !== pin.commit || gitText(directory, 'rev-parse', `refs/tags/${pin.tag}^{commit}`) !== pin.commit) throw new Error(`${repo.name}: checkout/tag differs from lock`);
    }
  }
  console.log('validate-family: ok');
} catch (error) { console.error(`validate-family: ${error.message}`); process.exitCode = 1; }

File: scripts/validate_family.py
#!/usr/bin/env python3
"""Validate portable metadata, locked inputs, and committed dependency boundaries."""
import argparse
from pathlib import Path
import re
import sys

from family import Family, FamilyError, git, read_toml


def validate_tree(path):
    # Inspect tracked inputs, never dependency caches, generated fusion trees,
    # or the deliberately adversarial conformance corpus.
    tracked = git(path, "ls-files", "-z", capture=True).split("\0")
    for relative in tracked:
        file = path / relative
        if not file.is_file() or relative.startswith("conformance/fixtures/"):
            continue
        if file.name == "Cargo.toml":
            data = read_toml(file)
            tables = [data]
            while tables:
                table = tables.pop()
                for value in table.values():
                    if not isinstance(value, dict):
                        continue
                    tables.append(value)
                    if "git" in value and ("branch" in value or not ("tag" in value or "rev" in value)):
                        raise FamilyError(f"{file}: Git dependencies require immutable tags/revisions")
                    if "path" in value:
                        target = (file.parent / value["path"]).resolve()
                        if not target.is_relative_to(path):
                            raise FamilyError(f"{file}: committed cross-repo path dependency")
        if relative.startswith(".github/workflows/") or relative == "action.yml":
            for use in re.findall(r"\buses:\s*[\"']?([^\s\"'#]+)", file.read_text()):
                if not use.startswith("./") and not re.fullmatch(r"[^@]+@[a-f0-9]{40}", use):
                    raise FamilyError(f"{file}: action must be pinned to a full commit SHA: {use}")


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--checkouts", action="store_true")
    args = parser.parse_args()
    family = Family(Path(__file__).resolve().parent.parent)
    if not (family.hub / "Cargo.lock").is_file():
        raise FamilyError("aggregate Cargo.lock is missing")
    read_toml(family.hub / "Cargo.lock")
    for repo in family.repos:
        path = family.path(repo)
        if not family.existing(repo):
            if args.checkouts:
                raise FamilyError(f"missing component: {repo['name']}")
            continue
        validate_tree(path)
        for required in ("AGENTS.md", "SPLIT.md", "agent/owner-map.json", "agent/test-map.json",
                         "agent/generated-zones.toml", "scripts/ci-local.sh", "ops/ci/required.sh"):
            if not (path / required).is_file():
                raise FamilyError(f"{repo['name']}: missing {required}")
        if args.checkouts and repo["name"] in family.pins:
            pin = family.pins[repo["name"]]
            if git(path, "rev-parse", "HEAD", capture=True) != pin["commit"]:
                raise FamilyError(f"{repo['name']}: HEAD differs from accepted lock")
            if git(path, "rev-parse", f"refs/tags/{pin['tag']}^{{commit}}", capture=True) != pin["commit"]:
                raise FamilyError(f"{repo['name']}: tag differs from accepted lock")
    print("validate-family: ok")


if __name__ == "__main__":
    try:
        main()
    except (FamilyError, OSError, ValueError) as exc:
        sys.exit(f"validate-family: {exc}")
